doc/developer_guide/developer_guide.rst

```diff
@@ -0,0 +1,341 @@
+Script-Languages-Container-Tool Developer Guide
+===============================================
+
+EXASLCT is the build tool for the script language container. This
+document is about the inner working of EXASLCT.
+
+Preparations
+------------
+
+This project uses `pre-commit <https://pre-commit.com/>`__ to run
+certain Githooks for validation. You don’t have to install
+``pre-commit`` as it will be installed with the ``exasol-toolbox``. You
+can activate the Githooks simply by running:
+
+.. code:: bash
+
+   pre-commit install
+
+However, if you don’t want to run all the checks during every commit,
+you can use:
+
+::
+
+   poetry run nox -s check
+
+to run all available checks on the project.
+
+About the Script Language Containers
+------------------------------------
+
+The Script Language Containers are getting build from several
+Dockerfiles which depend on each other. These Dockerfiles need to
+install all necessary dependencies for the `script
+client <https://github.com/exasol/script-languages/tree/master/exaudfclient/base>`__,
+compile the script client and install all necessary dependencies for the
+flavor and the customizations of the user.
+
+Problem Statement
+-----------------
+
+The old-style to build the containers was slow and laborious. A single
+change in the dependencies required a rebuild of everything.
+Furthermore, essential dependencies for the script client were mixed
+with flavor dependent dependencies. Changes were difficult for users and
+could break the container. It was actual unclear which of the
+dependencies were essential and which were not. For some flavors it was
+impossible to run the build on travis, because it exceeded the maximum
+runtime per job of 50 minutes. The build system and the test runner were
+both bash scripts which were messy and difficult to maintain. They
+worked with background jobs to do things in parallel which convoluted
+the logs which made error analysis difficult. Further, the test runner
+left temporary files which were owned by root, because they were created
+by a docker container.
+
+Design Goals
+------------
+
+- Easy customization of existing flavors for the user
+- User customizations are not able to break the container
+- Allowing different build steps per flavor
+- Separation of concern for code and Dockerfiles
+- Parallel builds
+- Partial builds
+- Local and remote caching
+- Faster development cycles for the script client
+- Allowing output redirection for testing of the flavor
+- Encapsulate running the tests and all its dependencies into docker
+  containers or volumes.
+- Error resilient
+
+Programming Model
+-----------------
+
+Exaslct is a mix of a build system, test runner and infrastructure as
```

Comment:
might be out of scope, but maybe double-check usage of Exaslct. There seems be three variants on spelling. I can see two (all caps vs all lower) of them perhaps referring to two distinct components, but idk.